content/en/blog/2023-01-10-building-a-raft-part-2.md

```diff
@@ -0,0 +1,100 @@
+
+So, when Beta and Gamma receive an `AppendEntriesRequest`, they must first ensure that Alpha's term not less than their own.
+
+
+#### 2. Reply false if log doesn’t contain an entry at prevLogIndex whose term matches prevLogTerm
+This step handles the problem of Log ordering. Consider the case where we have three sequential `AppendEntriesRequests` sent from Alpha to Beta and Gamma. Because of network delays, these may end up arriving at Gamma out of order, causing incorrect log ordering in Gamma’s log.
+
+!["Example of network delay"](/img/blog/2023-01-10-building-a-raft-part-2/networkDelay.png)
+
+To prevent this, Raft uses two fields in the `AppendEntriesRequest` containing the previous request the leader issued. In this way we can ensure that each follower has received the previous command before committing a new one. This also provides two emergent properties:
+
+- If two entries in different logs have the same index and term, then they store the same command
+- If two entries in different logs have the same index and term, then the logs are identical in all preceding entries
+
+Together these are called the Log Matching Property, and they are very important to Raft's safety and fault tolerance guarantees.
+
+#### 3. If an existing entry conflicts with a new one, delete the existing entry and all that follow it
+In the case of things like network partitions (let’s say one peer loses contact with the rest of the peer group, failing to sync with them, but is still receiving client requests), we may run into a situation where a server’s log is out of sync with the rest of the cluster.
+
+!["Possible follower states"](/img/blog/2023-01-10-building-a-raft-part-2/possibleFollowers.png)
+
+Algorithmically speaking, Beta will know it has identified a potential problem if Alpha's `AppendEntriesRequest` refers to a previous index for which Beta already has an entry with a term older than Alpha's term.
+
+#### 4. Append any new entries not already in the log.
+Following up from the previous step, if Beta finds any inconsistencies, it will overwrite it's log using the information included in Alpha's `AppendEntriesRequest`.
+
+#### 5. If leaderCommit > commitIndex, fast forward
+Finally, if Beta sees that Alpha's highest committed entry (`leaderCommit`) is greater than it's own local highest committed entry (`commitIndex`), Beta knows it needs to increase it's local index position. If Beta is still awaiting additional entries (e.g. from an `AppendEntriesRequest` it has received out of order), it will update its local `commitIndex` to be equal to the current length of the log. Otherwise, if Alpha's `AppendEntriesRequest` contains all of the latest entries Beta needs to be in sync, Beta will update it's local `commitIndex` to be the same as Alpha's.
```

Comment:
```suggestion
Finally, if Beta sees that Alpha's highest committed entry (`leaderCommit`) is greater than its own local highest committed entry (`commitIndex`), Beta knows it needs to increase its local index position. If Beta is still awaiting additional entries (e.g. from an `AppendEntriesRequest` it has received out of order), it will update its local `commitIndex` to be equal to the current length of the log. Otherwise, if Alpha's `AppendEntriesRequest` contains all of the latest entries Beta needs to be in sync, Beta will update its local `commitIndex` to be the same as Alpha's.
```